jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4296,36 +4296,108 @@ ConstantPool getConstantPool() {
+	 *    - the caller's class loader is not the same or an ancestor of the returned class
+	 *    - s.checkPackageAccess() disallows access to the package of the returned class
+	 * then throw a SecurityException.
+	 */
+	if (nestHost != this) {
+		SecurityManager securityManager = System.getSecurityManager();
+		if (nestHost == null) {
+			nestHost = this;
+		} else if (securityManager != null) {
+			ClassLoader callerClassLoader = ClassLoader.getCallerClassLoader();
+			ClassLoader nestHostClassLoader = nestHost.getClassLoader();			
+			if (!doesClassLoaderDescendFrom(nestHostClassLoader, callerClassLoader)) {
+				String nestHostPackageName = nestHost.getPackageName();
+				if (nestHostPackageName != null) {
+					securityManager.checkPackageAccess(nestHostPackageName);
+				}
+			}
+		}
+	}
+
+	return nestHost;
 }
 
 /**
  * Returns true if the class passed has the same nest top as this class.
- * The list of nest members in the classfile is permitted to contain
- * duplicates, or to explicitly include the nest host. It is not required
- * that an implementation of this method removes these duplicates.
  * 
  * @param		that		The class to compare
  * @return		true if class is a nestmate of this class; false otherwise.
  *
  */
 public boolean isNestmateOf(Class<?> that) {
-	return this.getNestHost() == that.getNestHost();
+	return this.getNestHostImpl() == that.getNestHostImpl();
 }
 
 /**
  * Answers the nest member classes of the receiver's nest host.
  *
+ * @throws		SecurityException if a SecurityManager is present & package access is not allowed
+ * @throws		LinkageError if there is any problem loading or validating a nest member or the nest host
  * @return		the host class of the receiver.
  */
-public Class<?>[] getNestMembers() {
-	return getNestMembersImpl();
+@CallerSensitive
+public Class<?>[] getNestMembers() throws LinkageError, SecurityException {
+	Class<?>[] nestMembers = getNestMembersImpl();
+	if (null != nestMembers) {
+		/* The specification requires that if:
+		 *    - a returned class is not the current class
+		 *    - a security manager, s, is present
+		 *    - the caller's class loader is not the same or an ancestor of the returned class
+		 *    - s.checkPackageAccess() disallows access to the package of the returned class
+		 * then throw a SecurityException. Since all classes in the same nest
+		 * must be in the same runtime package and thus the same class loader,
+		 * it is sufficient to check classloader ancestry once.
+		 */
+		SecurityManager securityManager = System.getSecurityManager();
+		if (null != securityManager) {
+			/* All classes in a nest must be in the same runtime package & therefore same classloader */
+			ClassLoader nestHostClassLoader = nestMembers[0].getClassLoader();
+			ClassLoader callerClassLoader = ClassLoader.getCallerClassLoader();
+			if (!doesClassLoaderDescendFrom(nestHostClassLoader, callerClassLoader)) {
+				for (Class<?> nm : nestMembers) {
```

Comment:
dont need to loop since they will all have the same package, do `this.getPackageName()`